DOC: Improve SIMD documentation(1/4) (#19581)
numpy/numpy: doc/source/reference/simd/how-it-works.rst

*****************************

> How the CPU dispatcher works?

Reviewer: ```suggestion
How does the CPU dispatcher work?
```

*****************************

> NumPy dispatcher is based on multi-source compiling, which means taking
> a certain source and compiling it multiple times with different compiler
> flags and also with different **C** definitions that affect the code
> paths to enable certain instruction-sets for each compiled object
> depending on the required optimizations, then combining the returned
> objects together.

Reviewer: This is a very long sentence - might be re-worded, maybe something like:
```suggestion
NumPy dispatcher is based on multi-source compiling, which means taking
a certain source and compiling it multiple times with different compiler
flags and also with different **C** definitions that affect the code
paths. This enables certain instruction-sets for each compiled object
depending on the required optimizations which are then combined in the returned
objects together.
```

Author: Done, but I have changed the last phrase to become "and ends with linking the
returned objects together.".

.. figure:: ../figures/opt-infra.png

This mechanism should support all compilers and it doesn't require any

> compiler-specific extension, but at the same time it is adds a few steps to
> normal compilation that are explained as follows:

Reviewer: ```suggestion
compiler-specific extension, but at the same time it adds a few steps to
normal compilation that are explained as follows.
```

1- Configuration
~~~~~~~~~~~~~~~~

Configuring the required optimization by the user before starting to build the
source files via the two command arguments as explained above:

-  ``--cpu-baseline``: minimal set of required optimizations.

-  ``--cpu-dispatch``: dispatched set of additional optimizations.

2- Discovering the environment
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In this part, we check the compiler and platform architecture
and cache some of the intermediary results to speed up rebuilding.

3- Validating the requested optimizations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

By testing them against the compiler, and seeing what the compiler can
support according to the requested optimizations.

4- Generating the main configuration header
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The generated header ``_cpu_dispatch.h`` contains all the definitions and
headers of instruction-sets for the required optimizations that have been
validated during the previous step.

It also contains extra C definitions that are used for defining NumPy's

> Python-level module attributes ``__cpu_baseline__`` and ``__cpu_dispaٍtch__``.

Reviewer: ```suggestion
Python-level module attributes ``__cpu_baseline__`` and ``__cpu_dispatch__``.
```

       * configuration header. These are definitions for the implied features.
       */
      #define NPY__CPU_TARGET_SSE
      #define NPY__CPU_TARGET_SSE2
      #define NPY__CPU_TARGET_SSE3
      #define NPY__CPU_TARGET_SSSE3
      #define NPY__CPU_TARGET_SSE41
      #define NPY__CPU_TARGET_POPCNT
      #define NPY__CPU_TARGET_SSE42
      #define NPY__CPU_TARGET_AVX
      #define NPY__CPU_TARGET_F16C
      #define NPY__CPU_TARGET_FMA3
      #define NPY__CPU_TARGET_AVX2
      #define NPY__CPU_TARGET_AVX512F
      // our dispatch-able source
      #include "/the/absuolate/path/of/hello.dispatch.c"

- **(D) Dispatch-able configuration header**: The infrastructure
  generates a config header for each dispatch-able source, this header
  mainly contains two abstract **C** macros used for identifying the
  generated objects, so they can be used for runtime dispatching
  certain symbols from the generated objects by any **C** source. It is
  also used for forward declarations.

  The generated header takes the name of the dispatch-able source after
  excluding the extension and replace it with '**.h**', for example

>   assume we have a dispatch-able source called **hello.dispatch.c** and

Reviewer: Because you were using double backticks for filenames before, I think it would be nice to keep a consistent formatting and use double backticks for all filenames.
